### PR TITLE
[HELM] Update template on port-foward usage guide NOTES.txt

### PR DESCRIPTION
**Pull request by phyyou**
- Fix Stale Docs URL in Helm Chart
- ~~Update kyuubiDefaults structure~~
- Update NOTES.txt
  - When kyuubi.kyuubiConf.kyuubiDefaults."kyuubi.frontend.bind.host" set localhost, render port-forward example  in NOTES.txt for development usage guide
- ~~Add extraFiles for config mount~~
  - ~~extraFiles for another file mount in kyuubi conf directory.~~

### _Why are the changes needed?_

Based on https://github.com/apache/kyuubi/issues/4561#issuecomment-1481026954_

> I think Helm chart notes had been wrong from this commit.

>https://github.com/apache/kyuubi/blob/dbb741fc5b0d2fd4b0640ff70ea12aca75864166/charts/kyuubi/templates/NOTES.txt#L33-L34

> Can we delete this note or update this note  if `--address` option is work like this? :
> ```suggestion
>  - To access {{ $.Release.Name }}-{{ $name | kebabcase }} service from outside the cluster for debugging, run the following > > 
> command: 
>     kubectl port-forward svc/{{ $.Release.Name }}-{{ $name | kebabcase }} {{ tpl $frontend.service.port $ }}:{{ tpl >?> >$frontend.service.port $ }} -n {{ $.Release.Namespace }} --address <your-ip>
> ```

~~and more changes:~~
- ~~Update `.Values.kyuubiConf.kyuubiDefaults` to dict for more variant usage in templates. (like NOTES.txt changes)~~
- ~~Add extraFiles in configmap for like mounting hive-site-xml on kyuubi config directory.~~ 

### _How was this patch tested?_

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by zwangsheng** [comment]
Hi @phyyou, thanks for your PR.

I install your changes in local minikube cluster, i find kyuubi register self to zookeeper with `localhost:10009`, which may cause engine can't find kyuubi instance.

```log
2023-03-24 02:56:27.968 INFO org.apache.kyuubi.ha.client.zookeeper.ZookeeperDiscoveryClient: Created a /kyuubi/serviceUri=localhost:10009;version=1.7.0;sequence=0000000000 on ZooKeeper for KyuubiServer uri: localhost:10009
```

You can find the corresponding logic in `TFrontendService.scala#connectionUrl`.

**Comment by pan3793** [comment]
> Update `.Values.kyuubiConf.kyuubiDefaults` to dict for more variant usage in templates. (like NOTES.txt changes)

I'm neutral on this change, @dnskr WDYT?

**Comment by dnskr** [comment]
> > Update `.Values.kyuubiConf.kyuubiDefaults` to dict for more variant usage in templates. (like NOTES.txt changes)
> 
> I'm neutral on this change, @dnskr WDYT?

It's against recommended best practices: https://helm.sh/docs/chart_best_practices/values/#naming-conventions
It looks natural in this case, but it brings many inconveniences with setting values, for example:
1. Copy/Paste `a.b.c` to IntelliJ IDEA becomes
```yaml
  a:
    b:
      c:
```
2. Set property value in cli: https://helm.sh/docs/chart_best_practices/values/#consider-how-users-will-use-your-values
```bash
helm install kyuubi --set kyuubiConf.kyuubiDefaults."kyuubi\.frontend\.bind\.host"=localhost
```
3. Overwriting many default values.

I know only one mature chart which still uses similar property names, but looks like it was a mistake they did at the beginning: https://github.com/argoproj/argo-helm/blob/ccef4448748601bb253a89fa86ed036b0e620cc1/charts/argo-cd/values.yaml#L136

**Comment by phyyou** [comment]
> It looks natural in this case, but it brings many inconveniences with setting values,

Thanks @dnskr. I definitely agree.
When I first thought of and created it, it seemed like a good idea, but when I actually tried Template Rendering with helm, and I realized it was inconvenient

I'd like to revert this, but I would like to think of a way to conveniently specify the configuration values in the Helm Chart.

Do you have any other good ideas for kyuubiConf in Helm Chart? @dnskr @pan3793

**Comment by phyyou** [comment]
I was just inspired by the Helm Chart of Apache Airflow (https://github.com/apache/airflow/blob/main/chart/values.yaml#L1834-L1909).

What do you think about creating kyuubiDefaults with a similar structure? @dnskr

File: charts/kyuubi/values.yaml [hunk 1]
```diff
@@ -101,17 +101,22 @@ server:
...
+  # See https://kyuubi.readthedocs.io/en/master/deployment/settings.html#kyuubi-configurations for more details
+  kyuubiDefaults: |
+
 
   # The value (templated string) is used for log4j2.xml file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#logging for more details
+  # See https://kyuubi.readthedocs.io/en/master/deployment/settings.html#logging for more details
   log4j2: ~
 
+  extraFiles: {}
```
Comment (pan3793) [inline charts/kyuubi/values.yaml:116]: it's valuable to change in a separate PR. let's focus on one thing in each PR, to make the future visitor easy to understand what we were doing.

Comment (phyyou) [reply charts/kyuubi/values.yaml:116]: Thanks! got it!

**Comment by pan3793** [comment]
> https://airflow.apache.org/docs/helm-chart/stable/parameters-ref.html

I like the idea of "examples", actually, I was asked, "I got no idea w/ a placeholder `~`, what should I assign to `kyuubiDefaults`?"

> ~
> ```
> Examples:
> 
> extraEnvFrom: |-
>   - secretRef:
>       name: '{{ .Release.Name }}-airflow-connections'
> extraEnvFrom: |-
>   - configMapRef:
>       name: '{{ .Release.Name }}-airflow-variables'
> ```

File: charts/kyuubi/values.yaml (hunk 1 above)
Comment (dnskr) [reply charts/kyuubi/values.yaml]: what is the use case for this property?

File: charts/kyuubi/values.yaml [hunk 2]
```diff
@@ -101,17 +101,19 @@ server:
...
   # The value (templated string) is used for kyuubi-env.sh file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#environments for more details
+  # See https://kyuubi.readthedocs.io/en/master/deployment/settings.html#environments for more details
   kyuubiEnv: ~
 
   # The value (templated string) is used for kyuubi-defaults.conf file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#kyuubi-configurations for more details
-  kyuubiDefaults: ~
+  # See https://kyuubi.readthedocs.io/en/master/deployment/settings.html#kyuubi-configurations for more details
+  kyuubiDefaults: |
```
Comment (pan3793) [inline charts/kyuubi/values.yaml:109]: it should be good to keep `~`, but let's add more comments, including "examples"

File: charts/kyuubi/values.yaml [hunk 3]
```diff
@@ -101,17 +101,19 @@ server:
...
+
 
   # The value (templated string) is used for log4j2.xml file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#logging for more details
+  # See https://kyuubi.readthedocs.io/en/master/deployment/settings.html#logging for more details
   log4j2: ~
 
+
```
Comment (pan3793) [inline charts/kyuubi/values.yaml:116]: nit: unnecessary change

File: charts/kyuubi/values.yaml (hunk 1 above)
Comment (pan3793) [reply charts/kyuubi/values.yaml]: Users may use Kyuubi/Spark w/ Hive MetaStore, then `hive-site.xml` is required

File: charts/kyuubi/values.yaml (hunk 3 above)
Comment (phyyou) [reply charts/kyuubi/values.yaml]: Thanks!

File: charts/kyuubi/values.yaml (hunk 1 above)
Comment (pan3793) [reply charts/kyuubi/values.yaml]: And another use case is `FileSessionConfAdvisor` introduced in #3742, and emm... it isn't mentioned in docs, and we appreciate contributors who can help us to improvement the docs 😄

**Comment by pan3793** [comment]
Do you have Kerberos use case? There are some extra work, like `keytab`, `krb5.conf`, to make the Helm chart work out-of-box

File: charts/kyuubi/values.yaml (hunk 1 above)
Comment (dnskr) [reply charts/kyuubi/values.yaml]: @pan3793  Would it be better to add specific properties instead? I.e. `hiveSite` etc

File: charts/kyuubi/values.yaml (hunk 2 above)
Comment (phyyou) [reply charts/kyuubi/values.yaml]: Thanks! I add Example usecases in kyuubiConf.

File: charts/kyuubi/values.yaml (hunk 1 above)
Comment (pan3793) [reply charts/kyuubi/values.yaml]: > @pan3793 Would it be better to add specific properties instead? I.e. `hiveSite` etc

I agree to use specific properties for specific files, but there are some non-enumerable files, e.g. in `FileSessionConfAdvisor` use case, there may have

- `kyuubi-session-etl.conf`
- `kyuubi-session-ad-hoc.conf`
- `kyuubi-session-admin.conf`
- `kyuubi-session-<...>.conf`

File: charts/kyuubi/values.yaml [hunk 4]
```diff
@@ -101,15 +101,33 @@ server:
...
+  #
+  # kyuubi.frontend.bind.host=10.0.0.1
+  #
+  # kyuubi.engine.type=SPARK_SQL
+  # kyuubi.engine.share.level=USER
+  # kyuubi.session.engine.initialize.timeout=PT3M
+  #
+  # kyuubi.ha.addresses=zk1:2181,zk2:2181,zk3:2181
+  # kyuubi.ha.namespace=kyuubi
+  #
```
Comment (pan3793) [inline charts/kyuubi/values.yaml:125]: I would suggest to including the "key" as well, user who not familiar w/ yaml will confusing how to handle `~`

```suggestion
  # Example:
  #
  #   kyuubiDefaults: |
  #     kyuubi.authentication=NONE
  #     kyuubi.frontend.bind.host=10.0.0.1
  #     kyuubi.engine.type=SPARK_SQL
  #     kyuubi.engine.share.level=USER
  #     kyuubi.session.engine.initialize.timeout=PT3M
  #     kyuubi.ha.addresses=zk1:2181,zk2:2181,zk3:2181
  #     kyuubi.ha.namespace=kyuubi
  #
```

Comment (phyyou) [reply charts/kyuubi/values.yaml:125]: Got it! Thanks for suggestion.

**Comment by dnskr** [comment]
> I just got a lot of inspiration Chart of Apache Airflow (https://github.com/apache/airflow/blob/main/chart/values.yaml#L1834-L1909).
> What do you think about kyuubiDefaults values with a similar structure? @dnskr

It might be a good idea, but managing many properties also might be exhausting. Also from my POV this requires stable configuration mechanism.

**Comment by phyyou** [comment]
> It might be a good idea, but managing many properties also might be exhausting. Also from my POV this requires stable configuration mechanism.

I agree with the previous comment that managing a large number of properties can be quite challenging and exhausting, especially if there is no stable configuration mechanism in place.

I think It meens while it could be a good long-term goal to move towards a more structured approach. I do not think that it is a good solution at the moment. 

So, I think that while a structured approach like the Apache Airflow chart could be useful in the future, it is not a good solution at the moment, and we should focus on creating a stable configuration mechanism first.

**Comment by phyyou** [comment]
> Do you have Kerberos use case? There are some extra work, like keytab, krb5.conf, to make the Helm chart work out-of-box

@pan3793

I don't have a complete understanding of the Kerberos use case, but I believe that we can refer to the configuration used in the Apache Airflow chart (https://github.com/apache/airflow/blob/main/chart/values.yaml#L413-L446) as a starting point.

So, I'd like to submit a new PR for krb5.conf with kyuubiConf.extraFiles.

File: charts/kyuubi/templates/NOTES.txt [hunk 5]
```diff
@@ -30,9 +30,13 @@ In order to check the release status, use:
 {{ $name | snakecase | upper }}:
 - To access {{ $.Release.Name }}-{{ $name | kebabcase }} service within the cluster, use the following URL:
     {{ $.Release.Name }}-{{ $name | kebabcase }}.{{ $.Release.Namespace }}.svc.cluster.local
+{{- if $.Values.kyuubiConf.kyuubiDefaults }}
+{{- if or (contains "kyuubi.frontend.bind.host=localhost" $.Values.kyuubiConf.kyuubiDefaults ) (contains "kyuubi.frontend.bind.host=127.0.0.1" $.Values.kyuubiConf.kyuubiDefaults) }}
```
Comment (pan3793) [inline charts/kyuubi/templates/NOTES.txt:34]: should be ok, even it allows
- `kyuubi.frontend.bind.host <any spaces> localhost`
- `kyuubi.frontend.bind.host <any spaces> = <any spaces> localhost`

**Review by pan3793 — APPROVED** [review]
LGTM except for minor comments, thanks

**Comment by pan3793** [comment]
@phyyou thanks for finding an example for kerberos, we can defer it if you don't have kerberos use case

File: charts/kyuubi/values.yaml [hunk 6]
```diff
@@ -101,15 +101,30 @@ server:
...
+  # export JAVA_HOME=/usr/jdk64/jdk1.8.0_152
+  # export SPARK_HOME=/opt/spark
+  # export FLINK_HOME=/opt/flink
+  # export HIVE_HOME=/opt/hive
+  # See example at conf/kyuubi-env.sh.template and https://kyuubi.readthedocs.io/en/master/deployment/settings.html#environments for more details
   kyuubiEnv: ~
 
   # The value (templated string) is used for kyuubi-defaults.conf file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#kyuubi-configurations for more details
+  # Example:
```
Comment (pan3793) [inline charts/kyuubi/values.yaml:113]: nit: new line before `Example:`

File: charts/kyuubi/values.yaml [hunk 7]
```diff
@@ -101,15 +101,30 @@ server:
 kyuubiConfDir: /opt/kyuubi/conf
 kyuubiConf:
   # The value (templated string) is used for kyuubi-env.sh file
-  # See https://kyuubi.apache.org/docs/latest/deployment/settings.html#environments for more details
+  # ## Examples ##
```
[the remark this answers is not in the document]
Comment (phyyou) [reply charts/kyuubi/values.yaml]: Thanks!

File: charts/kyuubi/templates/NOTES.txt (hunk 5 above)
Comment (phyyou) [reply charts/kyuubi/templates/NOTES.txt]: Got it.
I updated the Helm template to use a single regexMatch function to match localhost or 127.0.0.1 with any number of spaces.

File: charts/kyuubi/values.yaml (hunk 6 above)
Comment (phyyou) [reply charts/kyuubi/values.yaml:116]: Thanks

**Review by pan3793 — APPROVED** [review]

**Comment by phyyou** [comment]
@pan3793

Sorry! It has erorr with regex.

I'm fixing it now.

**Comment by phyyou** [comment]
@pan3793 

I'm sorry for the frequent mentions,

I fixed the when there was no equal sign now.

**Comment by zwangsheng** [comment]
> Sorry for the late response. I have also experienced this error before. Specifically, I encountered this error when set `kyuubi.frontend.connection.url.use.hostname=false.`, but, it may require a deeper investigation to identify the root cause.

Well, seems we revert  set `kyuubi.frontend.bind.host=localhost,` which cause the problem.

LGTM, thanks.

**Review by zwangsheng — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks all, merged to master/1.7

**Comment by phyyou** [comment]
@zwangsheng @pan3793 

Since [this commit](github.com/apache/kyuubi/commit/dbb741fc5b0d2fd4b0640ff70ea12aca75864166), 
`kyuubi.frontend.connection.url.use.hostname=false` is default value for helm chart.

In my opinion, it would be better not to use local Zookeeper in the Helm chart. For development environments.
I think it seems like a good idea to have an extra deployment option to turn on and off the Zookeeper deployment in the Helm chart.

**Comment by pan3793** [comment]
> For development environments, I think it seems like a good idea to have an extra deployment option to turn on and off the Zookeeper deployment in the Helm chart.

SGTM.

File: charts/kyuubi/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+ -->
+
+# Helm Chart for Apache Kyuubi
+
+[Apache Kyuubi](https://airflow.apache.org/) is a distributed and multi-tenant gateway to provide serverless SQL on Data Warehouses and Lakehouses.
```
Comment (yaooqinn) [inline charts/kyuubi/README.md:22]: There is a typo here https://airflow.apache.org/ cc @phyyou @zwangsheng @pan3793 @dnskr

Comment (phyyou) [reply charts/kyuubi/README.md:22]: Thanks
